scalyr_agent/monitor_utils/k8s.py

```diff
@@ -566,42 +580,209 @@ def process_object( self, obj ):
+        k8s = old_k8s
+        if k8s is None or old_config.need_new_k8s_object( new_config ):
+            changed = True
+            if new_config.verify_api_queries:
+                k8s = KubernetesApi(k8s_api_url=new_config.api_url)
+            else:
+                k8s = KubernetesApi( ca_file=None, k8s_api_url=new_config.k8s_api_url)
+
+        # create new filters if we need them
+        namespace_filter = old_namespace_filter
+        node_filter = old_node_filter
+        if old_config.need_new_filters( new_config ):
+            changed = True
+            namespace_filter = self._build_namespace_filter( new_config.namespaces_to_ignore )
+            node_filter = self._build_node_filter( k8s, namespace_filter )
+
+        #see if expiry or purge times have changed
+        if old_config.cache_expiry_secs != new_config.cache_expiry_secs or old_config.cache_purge_secs != new_config.cache_purge_secs:
+            changed = True
+
+        # update with new values
+        self._lock.acquire()
+        try:
+            # if config update count has changed, it means a newer config change
+            # came through on another thread before we finished this call, so only make our changes
+            # if the update counts are the same
+            if self._config_update_count != old_config_update_count:
+                changed = False
+            else:
 
-        self._thread.start()
+                self._k8s = k8s
+
+                self._namespace_filter = namespace_filter
+                self._node_filter = node_filter
+
+                self._cache_expiry_secs = new_config.cache_expiry_secs
+                self._cache_purge_secs = new_config.cache_expiry_secs
+
+                self._config = new_config
+
+                if changed:
+                    global_log.log( scalyr_logging.DEBUG_LEVEL_1, "Got new config %s", str( self._config ) )
+
+                self._config_update_count += 1
+        finally:
+            self._lock.release()
+
+        return changed
 
     def stop(self):
         """Stops the cache, specifically stopping the background thread that refreshes the cache"""
         self._thread.stop()
 
+    def start( self ):
+        """
+            Starts the background thread that reads from the k8s cache
+        """
+        if self._thread is None:
+            self._thread = StoppableThread( target=self.update_cache, name="K8S Cache" )
+            self._thread.start()
+
+    def update_config( self, new_config ):
+        """
+        Updates the cache config
+        """
+        old_config = _CacheConfig( api_url='', namespaces_to_ignore=[] )
+        self._lock.acquire()
+        try:
+            old_config = self._config
+        finally:
+            self._lock.release()
+
+        changed = self._configure( old_config=old_config, new_config=new_config )
+
+        if changed:
+            global_log.info( "K8s Cache configuration has been updated" )
```

Comment:
Maybe this should be a debug log?  I don't know if we really need to log this in normal cases.

Reply:
That being the case, the only reason we were returning True or False in the _configure method was to output an update.  I've therefore gone through and updated the previous changes to no longer return True/False and just keep the debug output in that function instead.

Reply:
That's fine... yes, you can remove the return value here and just keep the debug version of this in the function itself.